File: opem/Dynamic/Padulles_Amphlett.py
# -*- coding: utf-8 -*-
from opem.Params import Padulles_Amphlett_InputParams as InputParams
from opem.Params import Padulles_Amphlett_Outparams as OutputParams
from opem.Params import Padulles_Amphlett_Params_Default as Defaults
from opem.Static.Amphlett import Power_Calc,Eta_Act_Calc,Eta_Conc_Calc,Eta_Ohmic_Calc,Loss_Calc
from opem.Dynamic.Padulles1 import PH2_Calc,PO2_Calc,Kr_Calc,qO2_Calc,Efficiency_Calc
from opem.Dynamic.Padulles2 import Enernst_Calc,PH2O_Calc
from opem.Dynamic.Padulles_Hauer import qH2_Calc
from opem.Functions import *
from opem.Params import Padulles_Amphlett_Description
import os

def Vcell_Calc(Enernst, Loss, N):
    """
    This function calculate cell voltage
    :param Enernst:  Enernst [V}
    :type Enernst : float
    :param Loss:  Loss [V]
    :type Loss : float
    :return:  Cell voltage [V] as float
    """
    try:
        result = Enernst - N*Loss
        return result
    except Exception:
        print("[Error] Vcell Calculation Error (Enernst:%s, Loss:%s, N:%s)"%(str(Enernst),str(Loss),str(N)))

def Dynamic_Analysis(InputMethod=Get_Input, TestMode=False, PrintMode=True, ReportMode=True):
    """
    This function run Padulles-Amphlett analysis  with calling other functions
    :param InputMethod : Input Function Or Input Test Vector
    :param TestMode : Test Mode Flag
    :type InputMethod : dict or Get_Input function object
    :type TestMode:bool
    :param PrintMode : Print Mode Control Flag (True : Print Outputs)
    :type PrintMode:bool
    :param ReportMode : Report Mode Control Flag (True : Generate Report)
    :type ReportMode: bool
    :return: Result as dict
    """
    OutputFile = None
    CSVFile = None
    Warning1 = False
    Warning2 = False
    I_Warning = 0
    try:
        Simulation_Title="Padulles-Amphlett"
        if PrintMode==True:
            print("###########")
            print(Simulation_Title+"-Model Simulation")
            print("###########")
        OutputParamsKeys = list(OutputParams.keys())
        OutputParamsKeys.sort()
        Output_Dict = dict(zip(OutputParamsKeys, [None] * len(OutputParamsKeys)))
        if not TestMode:
            Input_Dict = InputMethod(InputParams,params_default=Defaults)
        else:
            Input_Dict = InputMethod
            Input_Dict = filter_default(input_dict=Input_Dict, params_default=Defaults)
        Input_Dict = filter_lambda(Input_Dict)
        if PrintMode==True:
            print("Analyzing . . .")
        Name = Input_Dict["Name"]
        if ReportMode==True:
            OutputFile = Output_Init(Input_Dict,Simulation_Title,Name)
            CSVFile = CSV_Init(OutputParamsKeys,OutputParams,Simulation_Title,Name)
            HTMLFile = HTML_Init(Simulation_Title, Name)
        IEndMax = Input_Dict["JMax"] * Input_Dict["A"]
        IEnd = min(IEndMax, Input_Dict["i-stop"])
        IEnd = Input_Dict["i-stop"]
        IStep = Input_Dict["i-step"]
        Precision = get_precision(IStep)
        i = Input_Dict["i-start"]
        I_List = []
        Power_List = []
        Vstack_List = []
        Kr=Kr_Calc(Input_Dict["N0"])
        qH2=qH2_Calc(Input_Dict["qMethanol"],Input_Dict["CV"],Input_Dict["t1"],Input_Dict["t2"])
        qO2=qO2_Calc(qH2,Input_Dict["rho"])
        while i < IEnd:
            try:
                I_List.append(i)
                Output_Dict["PO2"] = PO2_Calc(Input_Dict["KO2"], Input_Dict["tO2"], Kr, i, qO2)
                Output_Dict["PH2"] = PH2_Calc(Input_Dict["KH2"], Input_Dict["tH2"], Kr, i, qH2)
                Output_Dict["Eta Activation"] = Eta_Act_Calc(Input_Dict["T"], Output_Dict["PO2"], Output_Dict["PH2"], i,
                                                             Input_Dict["A"])
                Output_Dict["Eta Ohmic"] = Eta_Ohmic_Calc(i, Input_Dict["l"], Input_Dict["A"], Input_Dict["T"],
                                                          Input_Dict["lambda"], R_elec=Input_Dict["R"])
                Output_Dict["Eta Concentration"] = Eta_Conc_Calc(i, Input_Dict["A"], Input_Dict["B"],
                                                                 Input_Dict["JMax"])
                Output_Dict["Loss"] = Loss_Calc(Output_Dict["Eta Activation"], Output_Dict["Eta Ohmic"],
                                                Output_Dict["Eta Concentration"])
                Output_Dict["PH2O"]=PH2O_Calc(Input_Dict["KH2O"],Input_Dict["tH2O"],Kr,i,qH2)
                Output_Dict["E"]=Enernst_Calc(Input_Dict["E0"],Input_Dict["N0"],Input_Dict["T"],Output_Dict["PH2"],Output_Dict["PO2"],Output_Dict["PH2O"])
                Output_Dict["FC Voltage"]=Vcell_Calc(Output_Dict["E"], Output_Dict["Loss"],Input_Dict["N0"])
                [Warning1, I_Warning] = warning_check_1(Output_Dict["FC Voltage"], I_Warning, i, Warning1)
                Warning2 = warning_check_2(Vcell=Output_Dict["FC Voltage"], warning_flag=Warning2)
                Vstack_List.append(Output_Dict["FC Voltage"])
                Output_Dict["FC Efficiency"] = Efficiency_Calc(Output_Dict["FC Voltage"],Input_Dict["N0"])
                Output_Dict["FC Power"] = Power_Calc(Output_Dict["FC Voltage"], i)
                Power_List.append(Output_Dict["FC Power"])
                if ReportMode==True:
                    Output_Save(OutputParamsKeys, Output_Dict,OutputParams, i, OutputFile,PrintMode)
                    CSV_Save(OutputParamsKeys, Output_Dict, i, CSVFile)
                i = rounder(i + IStep, Precision)
            except Exception as e:
                print(str(e))
                i = rounder(i + IStep, Precision)
                if ReportMode==True:
                    Output_Save(OutputParamsKeys, Output_Dict, OutputParams, i, OutputFile,PrintMode)
                    CSV_Save(OutputParamsKeys, Output_Dict, i, CSVFile)
        if ReportMode==True:
            HTML_Desc(Simulation_Title, Padulles_Amphlett_Description, HTMLFile)
            HTML_Input_Table(Input_Dict=Input_Dict, Input_Params=InputParams, file=HTMLFile)
            HTML_Chart(x=str(I_List), y=str(Power_List), color='rgba(255,99,132,1)', x_label="I(A)", y_label="P(W)",
                    chart_name="FC-Power", size="600px", file=HTMLFile)
            HTML_Chart(x=str(I_List), y=str(Vstack_List), color='rgba(99,100,255,1)', x_label="I(A)", y_label="V(V)",
                    chart_name="FC-Voltage", size="600px", file=HTMLFile)
            warning_print(warning_flag_1=Warning1, warning_flag_2=Warning2, I_Warning=I_Warning, file=HTMLFile,
                          PrintMode=PrintMode)
            HTML_End(HTMLFile)
            OutputFile.close()
            CSVFile.close()
            HTMLFile.close()
        if PrintMode==True:
            print("Done!")
        if not TestMode:
            if PrintMode==True:
                print("Result In -->" + os.path.join(os.getcwd(),Simulation_Title))
        else:
            return {"P": Power_List, "I": I_List, "V": Vstack_List}
    except Exception:
        print("[Error] Padulles-Amphlett Dynamic Simulation Failed!(Check Your Inputs)")

File: MATLAB/Padulles_Amphlett.m
close all

try
    python_version = pyversion;
    fprintf(2,'** Python Version : %s\n',python_version);
catch e
    fprintf(2,'** Error : %s\n',e.message);
end

% Import model
opem = py.importlib.import_module('opem');
model = opem.Dynamic.Padulles_Amphlett;
test_vector = opem.Params.Padulles_Amphlett_Standard_Vector;

% Model inputs
test_vector{'A'} = 50.6; % Active area [cm^2]
test_vector{'l'} = 0.0178; % Membrane thickness [cm]
test_vector{'lambda'} = 23; % An adjustable parameter with a min value of 14 and max value of 23
test_vector{'JMax'} = 1.5; % Maximum current density [A/(cm^2)]
test_vector{'T'} = 343; % Fuel cell temperature [K]
test_vector{'N0'} = 5; % No load voltage [V]
test_vector{'KO2'} = 0.0000211; % Oxygen valve constant [kmol.s^(-1).atm^(-1)]
test_vector{'KH2'} = 0.0000422; % Hydrogen valve constant [kmol.s^(-1).atm^(-1)]
test_vector{'KH2O'} = 0.000007716; % Water valve constant [kmol.s^(-1).atm^(-1)]
test_vector{'tH2'} = 3.37; % Hydrogen time constant [s]
test_vector{'tO2'} = 6.74; % Oxygen time constant [s]
test_vector{'t1'} = 2; % Reformer time constant [s]
test_vector{'t2'} = 2; % Reformer time constant [s]
test_vector{'tH2O'} = 18.418; % Water time constant [s]
test_vector{'rho'} = 1.168; % Hydrogen-Oxygen flow rate
test_vector{'qMethanol'} = 0.0002; % Molar flow of methanol [kmol.s^(-1)]
test_vector{'CV'} = 2; % Conversion factor
test_vector{'i-start'} = 0.1; % Cell operating current start point [A]
test_vector{'i-stop'} = 75; % Cell operating current end point [A]
test_vector{'i-step'} = 0.1; % Cell operating current step
test_vector{'Name'} = 'Padulles_Amphlett_Test';

% Run simulation
test_mode = true;
print_mode = true;
report_mode = false;
result = model.Dynamic_Analysis(test_vector,test_mode,print_mode,report_mode);

% Model outputs
P = cellfun(@vector_filter, cell(result{'P'}));
I = cellfun(@vector_filter, cell(result{'I'}));
V = cellfun(@vector_filter, cell(result{'V'}));
EFF = cellfun(@vector_filter, cell(result{'EFF'}));
Ph = cellfun(@vector_filter, cell(result{'Ph'}));
VE = cellfun(@vector_filter, cell(result{'VE'}));
PO2 = cellfun(@vector_filter, cell(result{'PO2'}));
PH2 = cellfun(@vector_filter, cell(result{'PH2'}));
PH2O = cellfun(@vector_filter, cell(result{'PH2O'}));
Eta_Active = cellfun(@vector_filter, cell(result{'Eta_Active'}));
Eta_Conc = cellfun(@vector_filter, cell(result{'Eta_Conc'}));
Eta_Ohmic = cellfun(@vector_filter, cell(result{'Eta_Ohmic'}));

% Power-Stack
figure(1)
plot(I,P)
xlabel('I(A)')
ylabel('P(W)')
legend('Power-Stack')

% Voltage-Stack & Linear Approximation
figure(2)
plot(I,V)
xlabel('I(A)')
ylabel('V(V)')
hold on
plot(I,VE)
legend('Voltage-Stack','Linear-Apx')

% Efficiency
figure(3)
plot(I,EFF)
xlabel('I(A)')
ylabel('EFF')
legend('Efficiency')

% Power-Thermal
figure(4)
plot(I,Ph)
xlabel('I(A)')
ylabel('P(W)')
legend('Power(Thermal)')

% PO2
figure(5)
plot(I,PO2)
xlabel('I(A)')
ylabel('PO2(atm)')
legend('PO2')

% PH2
figure(6)
plot(I,PH2)
xlabel('I(A)')
ylabel('PH2(atm)')
legend('PH2')

% PH2O
figure(7)
plot(I,PH2O)
xlabel('I(A)')
ylabel('PH2O(atm)')
legend('PH2O')

% Eta-Active , Eta-Conc and Eta-Ohmic
figure(8)
plot(I,Eta_Active)
xlabel('I(A)')
ylabel('V(V)')
hold on
plot(I,Eta_Conc)
hold on
plot(I,Eta_Ohmic)
legend('Eta Active','Eta Conc','Eta Ohmic')

